File: example/iTEBD/iTEBD.cpp
#include "cytnx.hpp"
#include <iostream>
#include <cmath>  // abs

using namespace std;
using namespace cytnx;

typedef cytnx::Accessor ac;

int main(int argc, char *argv[]) {
  unsigned int chi = 20;
  double J = 1.0;
  double Hx = 1.0;
  double CvgCrit = 1.0e-10;
  double dt = 0.1;

  //> Create onsite-Op
  Tensor Sz = cytnx::zeros({2, 2});
  Sz.at<double>({0, 0}) = 1;
  Sz.at<double>({1, 1}) = -1;

  Tensor Sx = cytnx::zeros({2, 2});
  Sx.at<double>({0, 1}) = Sx.at<double>({1, 0}) = Hx;
  Tensor I = Sz.clone();
  I.at<double>({1, 1}) = 1;

  // cout << Sz << Sx << I << endl;

  //> Build Evolution Operator
  Tensor TFterm = cytnx::linalg::Kron(Sx, I) + cytnx::linalg::Kron(I, Sx);
  Tensor ZZterm = cytnx::linalg::Kron(Sz, Sz);
  Tensor tH = Hx * TFterm + J * ZZterm;

  Tensor teH = cytnx::linalg::ExpH(tH, -dt);
  teH.reshape_({2, 2, 2, 2});
  cout << teH;
  tH.reshape_({2, 2, 2, 2});

  UniTensor eH = UniTensor(teH, 2);
  eH.print_diagram();
  cout << eH;

  UniTensor H = UniTensor(tH, 2);
  H.print_diagram();

  //> Create MPS:
  //
  //     |    |
  //   --A-la-B-lb--
  //
  UniTensor A = UniTensor({cyx::Bond(chi), cyx::Bond(2), cyx::Bond(chi)}, {-1, 0, -2}, 1);
  UniTensor B = UniTensor(A.bonds(), {-3, 1, -4}, 1);

  cytnx::random::Make_normal(B.get_block_(), 0, 0.2);
  cytnx::random::Make_normal(A.get_block_(), 0, 0.2);

  A.print_diagram();
  B.print_diagram();

  UniTensor la =
    UniTensor({cyx::Bond(chi), cyx::Bond(chi)}, {-2, -3}, 1, Type.Double, Device.cpu, true);
  UniTensor lb =
    UniTensor({cyx::Bond(chi), cyx::Bond(chi)}, {-4, -5}, 1, Type.Double, Device.cpu, true);
  la.put_block(cytnx::ones(chi));
  lb.put_block(cytnx::ones(chi));

  la.print_diagram();
  lb.print_diagram();
  //> Evov:
  double Elast = 0;

  for (unsigned int i = 0; i < 10000; i++) {
    A.set_labels({-1, 0, -2});
    B.set_labels({-3, 1, -4});
    la.set_labels({-2, -3});
    lb.set_labels({-4, -5});

    // contract all
    UniTensor X = cyx::Contract(cyx::Contract(A, la), cyx::Contract(B, lb));
    lb.set_label(1, -1);
    X = cyx::Contract(lb, X);

    // X =
    //           (0)  (1)
    //            |    |
    //  (-4) --lb-A-la-B-lb-- (-5)
    //
    UniTensor Xt = X.clone();

    //> calculate norm and energy for this step
    // Note that X,Xt contract will result a rank-0 tensor, which can use item() toget element
    double XNorm = cyx::Contract(X, Xt).item<double>();
    UniTensor XH = cyx::Contract(X, H);

    XH.set_labels({-4, -5, 0, 1});
    double XHX = cyx::Contract(Xt, XH).item<double>();
    double E = XHX / XNorm;

    //> check if converged.
    if (abs(E - Elast) < CvgCrit) {
      cout << "[Converged!]" << endl;
      break;
    }
    cout << "Step: " << i << "Enr: " << Elast << endl;
    Elast = E;

    //> Time evolution the MPS
    UniTensor XeH = cyx::Contract(X, eH);
    XeH.permute_({-4, 2, 3, -5}, -1, true);

    //> Do Svd + truncate
    //
    //        (2)   (3)                   (2)                                    (3)
    //         |     |          =>         |         +   (-6)--s--(-7)  +         |
    //  (-4) --= XeH =-- (-5)        (-4)--U--(-6)                          (-7)--Vt--(-5)
    //
    XeH.set_Rowrank(2);
    vector<UniTensor> out = cyx::xlinalg::Svd_truncate(XeH, chi);
    la = out[0];
    A = out[1];
    B = out[2];
    double Norm = cytnx::linalg::Norm(la.get_block_()).item<double>();
    la *= 1. / Norm;  // normalize

    // de-contract the lb tensor , so it returns to
    //
    //            |     |
    //       --lb-A'-la-B'-lb--
    //
    // again, but A' and B' are updated
    A.set_labels({-1, 0, -2});
    A.set_Rowrank(1);
    B.set_labels({-3, 1, -4});
    B.set_Rowrank(1);

    UniTensor lb_inv = 1. / lb;
    A = cyx::Contract(lb_inv, A);
    B = cyx::Contract(B, lb_inv);

    //> translation symm, exchange A and B site
    UniTensor tmp = A;
    A = B;
    B = tmp;

    tmp = la;
    la = lb;
    lb = tmp;
  }

  return 0;
}

File: example/iTEBD/iTEBD.py
import cytnx
import numpy as np

##
# Author: Kai-Hsin Wu
##

#Example of 1D Ising model
## iTEBD
##-------------------------------------

def itebd_tfim(chi = 20, J = 1.0, Hx = 1.0, dt = 0.1, CvgCrit = 1.0e-10):

    ## Create onsite-Op
    Sz = cytnx.physics.pauli("z").real()
    Sx = cytnx.physics.pauli("x").real()
    I = cytnx.eye(2)

    ## Build Evolution Operator
    TFterm = cytnx.linalg.Kron(Sx,I) + cytnx.linalg.Kron(I,Sx)
    ZZterm = cytnx.linalg.Kron(Sz,Sz)

    H = Hx*TFterm + J*ZZterm
    del TFterm, ZZterm
    print(H)

    eH = cytnx.linalg.ExpH(H,-dt) ## or equivantly ExpH(-dt*H)
    eH.reshape_(2,2,2,2)
    print(eH)
    H.reshape_(2,2,2,2)

    eH = cytnx.UniTensor(eH)
    eH.print_diagram()
    print(eH)

    H = cytnx.UniTensor(H)
    H.print_diagram()


    ## Create MPS:
    #
    #     |    |
    #   --A-la-B-lb--
    #
    A = cytnx.UniTensor([cytnx.Bond(chi),cytnx.Bond(2),cytnx.Bond(chi)],labels=['a','0','b']);
    B = cytnx.UniTensor(A.bonds(),rowrank=1,labels=['c','1','d']);
    cytnx.random.normal_(B.get_block_(),0,0.2);
    cytnx.random.normal_(A.get_block_(),0,0.2);
    A.print_diagram()
    B.print_diagram()
    #print(A)
    #print(B)

    la = cytnx.UniTensor([cytnx.Bond(chi),cytnx.Bond(chi)],labels=['b','c'],is_diag=True)
    lb = cytnx.UniTensor([cytnx.Bond(chi),cytnx.Bond(chi)],labels=['d','e'],is_diag=True)
    la.put_block(cytnx.ones(chi));
    lb.put_block(cytnx.ones(chi));
    la.print_diagram()
    lb.print_diagram()
    #print(la)
    #print(lb)

    ## Evov:
    Elast = 0
    for i in range(10000):

        A.set_labels(['a','0','b'])
        B.set_labels(['c','1','d'])
        la.set_labels(['b','c'])
        lb.set_labels(['d','e'])


        ## contract all
        X = cytnx.Contract(cytnx.Contract(A,la),cytnx.Contract(B,lb))
        lb_l = lb.relabel("e","a")
        X = cytnx.Contract(lb_l,X)


        ## X =
        #           (0)  (1)
        #            |    |
        #  (d) --lb-A-la-B-lb-- (e)
        #
        # X.print_diagram()
        Xt = X.clone()


        ## calculate norm and energy for this step
        # Note that X,Xt contract will result a rank-0 tensor, which can use item() toget element
        XNorm = cytnx.Contract(X,Xt).item()
        XH = cytnx.Contract(X,H)
        XH.set_labels(['d','e','0','1'])


        XHX = cytnx.Contract(Xt,XH).item() ## rank-0
        E = XHX/XNorm

        # print(E)
        ## check if converged.
        if(np.abs(E-Elast) < CvgCrit):
            print("[Converged!]")
            break
        print("Step: %d Enr: %5.8f"%(i,Elast))
        Elast = E


        ## Time evolution the MPS
        XeH = cytnx.Contract(X,eH)
        XeH.permute_(['d','2','3','e'])
        # XeH.print_diagram()

        ## Do Svd + truncate
        ##
        #        (2)   (3)                   (2)                                               (3)
        #         |     |          =>         |            + (_aux_L)--s--(_aux_R)  +           |
        #  (d) --= XeH =-- (e)           (d)--U--(_aux_L)                             (_aux_R)--Vt--(e)
        #

        XeH.set_rowrank_(2)
        la,A,B = cytnx.linalg.Svd_truncate(XeH,chi)
        la.normalize_()

        #A.print_diagram()
        #la.print_diagram()
        #B.print_diagram()


        # de-contract the lb tensor , so it returns to
        #             2     3
        #             |     |
        #       d--lb-A'-la-B'-lb--e
        #
        # again, but A' and B' are updated
        lb_inv = 1./lb
        # lb_inv.print_diagram();
        lb_inv.set_labels(['e','d'])

        A = cytnx.Contract(lb_inv,A)
        B = cytnx.Contract(B,lb_inv)

        # translation symmetry, exchange A and B site
        A,B = B,A
        la,lb = lb,la
    return Elast

if __name__ == '__main__':
    itebd_tfim()
